app/assets/javascripts/discourse/app/components/d-multi-select.gjs

```diff
@@ -0,0 +1,256 @@
+
+class Skeleton extends Component {
+  get width() {
+    return htmlSafe(`width: ${Math.floor(Math.random() * 70) + 20}%`);
+  }
+
+  <template>
+    <div class="d-multi-select__skeleton">
+      <div class="d-multi-select__skeleton-checkbox" />
+      <div class="d-multi-select__skeleton-text" style={{this.width}} />
+    </div>
+  </template>
+}
+
+export default class DMultiSelect extends Component {
+  @tracked searchTerm = "";
+
+  @tracked preselectedItem = null;
+
+  @tracked selection = new TrackedArray(this.args.selection);
+
+  @tracked results = null;
+
+  get hasSelection() {
+    return this.selection.length > 0;
+  }
+
+  get label() {
+    return this.args.label ?? i18n("multi_select.label");
+  }
+
+  @action
+  resultsChanged(results) {
+    this.preselectedItem = null;
+    this.results = results;
+  }
+
+  @action
+  search(event) {
+    this.searchTerm = event.target.value;
+  }
+
+  @action
+  focus(input) {
+    input.focus();
+  }
+
+  @action
+  handleKeydown(event) {
+    if (event.key === "Enter") {
+      event.preventDefault();
+
+      if (this.preselectedItem) {
+        this.toggle(this.preselectedItem);
+      }
+    }
+
+    if (event.key === "ArrowDown") {
+      event.preventDefault();
+
+      if (!this.results || this.results.length === 0) {
+        return;
+      }
+
+      if (this.preselectedItem === null) {
+        this.preselectedItem = this.results[0];
+      } else {
+        const currentIndex = this.results.findIndex((item) =>
+          this.compare(item, this.preselectedItem)
+        );
+
+        if (currentIndex < this.results.length - 1) {
+          this.preselectedItem = this.results[currentIndex + 1];
+        }
+      }
+    }
+
+    if (event.key === "ArrowUp") {
+      event.preventDefault();
+
+      if (!this.results || this.results.length === 0) {
+        return;
+      }
+
+      if (this.preselectedItem === null) {
+        this.preselectedItem = this.results[0];
+      } else {
+        const currentIndex = this.results.findIndex((item) =>
+          this.compare(item, this.preselectedItem)
+        );
+
+        if (currentIndex > 0) {
+          this.preselectedItem = this.results[currentIndex - 1];
+        }
+      }
+    }
+  }
+
+  @action
+  remove(selectedItem, event) {
+    event.stopPropagation();
+
+    this.selection = new TrackedArray(
+      this.selection.filter((item) => !this.compare(item, selectedItem))
+    );
+  }
+
+  @action
+  isSelected(result) {
+    return this.selection.filter((item) => this.compare(item, result)).length;
+  }
+
+  @action
+  toggle(result, event) {
+    event?.stopPropagation();
+
+    if (this.isSelected(result)) {
+      this.selection = new TrackedArray(
+        this.selection.filter((item) => !this.compare(item, result))
+      );
+    } else {
+      this.selection.push(result);
+    }
+
+    this.args.onChange?.(this.selection);
+  }
+
+  @action
+  compare(a, b) {
+    return this.args.compareFn?.(a, b) ?? a.id === b.id;
```

Comment:
What if `compareFn` exists, but returns a nullish value? I feel like this kind of thing might be better?

```suggestion
    if(this.args.compareFn){
      return this.args.compareFn(a, b);
    } else {
      return a.id === b.id;
    }
```

Reply:
Yes not against it :+1: